src/mprk.jl

```diff
@@ -24,19 +24,29 @@ function build_mprk_matrix(P, sigma, dt)
 end
 
 # in-place for dense arrays
-function build_mprk_matrix!(M, P, sigma, dt)
+function build_mprk_matrix!(M, P, sigma, dt; d = nothing)
     # M[i,i] = (sigma[i] + dt * sum_j P[j,i]) / sigma[i]
     # M[i,j] = -dt * P[i,j] / sigma[j]
     # TODO: the performance of this can likely be improved
     Base.require_one_based_indexing(M, P, sigma)
     @assert size(M, 1) == size(M, 2) == size(P, 1) == size(P, 2) == length(sigma)
+    if !isnothing(d)
+        Base.require_one_based_indexing(d)
+        @assert length(sigma) == length(d)
+    end
 
     zeroM = zero(eltype(P))
 
     # Set sigma on diagonal
     @inbounds for i in eachindex(sigma)
         M[i, i] = sigma[i]
     end
+    # Add nonconservative destruction terms to diagonal (PDSFunctions only!)
+    if !isnothing(d)
+        @inbounds for i in eachindex(d)
+            M[i, i] += dt * d[i]
+        end
+    end
```

Comment:
Although it will likely not be a bottleneck, we could consider improving the performance a bit by adding `sigma` and `dt * d` together, e.g.,
```suggestion
    if isnothing(d)
        # Set only sigma on diagonal
        @inbounds for i in eachindex(sigma)
            M[i, i] = sigma[i]
        end
    else
        # Set sigma and nonconservative destruction terms 
        # on diagonal (for non-conservative PDSFunctions only!)
        @inbounds for i in eachindex(d)
            M[i, i] = sigma[i] + dt * d[i]
        end
    end
```

But I don't have a strong opinion on this - maybe we should benchmark it?

Reply:
Don't see any difference.

Old implementation:

```julia
julia> @benchmark solve($prob_pds_stratreac, $MPE(), dt = 10.0)
BenchmarkTools.Trial: 252 samples with 1 evaluation.
 Range (min … max):  17.963 ms … 26.466 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.198 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   19.870 ms ±  1.966 ms  ┊ GC (mean ± σ):  2.94% ± 5.59%

  █ ▂▁▁▁    ▁                                                  
  █▅█████▆███▆▅▅▄▃▄▄▂▂▄▂▄▄▄▃▄▃▃▃▃▃▂▃▂▄▃▂▁▄▁▄▃▂▂▁▄▃▂▁▂▁▂▁▃▁▁▂▂ ▃
  18 ms           Histogram: frequency by time        25.7 ms <

 Memory estimate: 11.87 MiB, allocs estimate: 51862.
```

New implementation:

```julia
 julia> @benchmark solve($prob_pds_stratreac, $MPE(), dt = 10.0)
BenchmarkTools.Trial: 251 samples with 1 evaluation.
 Range (min … max):  17.934 ms … 26.443 ms  ┊ GC (min … max): 0.00% … 19.74%
 Time  (median):     19.061 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   19.911 ms ±  2.061 ms  ┊ GC (mean ± σ):  2.86% ±  5.47%

  █     ▁                                                      
  █▃▇▅▆▅█▆▆▆▅▄▄▂▃▄▃▃▃▃▃▁▃▃▃▃▂▃▃▃▂▃▃▁▃▃▂▃▃▁▂▂▃▂▃▃▃▃▁▂▁▁▃▂▁▂▁▁▂ ▃
  17.9 ms         Histogram: frequency by time        25.9 ms <

 Memory estimate: 11.87 MiB, allocs estimate: 51862.
```

Reply:
Thanks! Let's keep the simpler one 👍

Reply:
:smile:  Which one is that?

Reply:
The one you have implemented, not my suggestion from above 🙂